projects/solar/src/main.c

```diff
@@ -1,2 +1,123 @@
-// Placeholder so the build doesn't fail
-int main(void) {}
+#include <stdbool.h>
+
+#include "adc.h"
+#include "can.h"
+#include "command_rx.h"
+#include "data_store.h"
+#include "drv120_relay.h"
+#include "event_queue.h"
+#include "fault_handler.h"
+#include "fault_monitor.h"
+#include "gpio.h"
+#include "i2c.h"
+#include "interrupt.h"
+#include "log.h"
+#include "logger.h"
+#include "mcp3427_adc.h"
+#include "mppt.h"
+#include "relay_fsm.h"
+#include "sense.h"
+#include "sense_mcp3427.h"
+#include "sense_mppt.h"
+#include "sense_temperature.h"
+#include "soft_timer.h"
+#include "solar_boards.h"
+#include "solar_config.h"
+#include "spi.h"
+#include "status.h"
+#include "wait.h"
+
+// TODO(SOFT-293): get hardware to solder something so we can detect MPPT counts instead of this
+#ifndef MPPT_COUNT
+#define MPPT_COUNT SOLAR_BOARD_6_MPPTS
+#endif
+
+#define SENSE_CYCLE_PERIOD_US 1000000  // 1 second
+
+static CanStorage s_can_storage;
+static RelayFsmStorage s_relay_fsm_storage = { 0 };
+
+static StatusCode prv_initialize_libraries(void) {
+  interrupt_init();
+  soft_timer_init();
+  event_queue_init();
+  status_ok_or_return(gpio_init());
+  adc_init(ADC_MODE_SINGLE);
+
+  status_ok_or_return(i2c_init(I2C_PORT_1, config_get_i2c1_settings()));
+  status_ok_or_return(i2c_init(I2C_PORT_2, config_get_i2c2_settings()));
+  status_ok_or_return(spi_init(SPI_PORT_2, config_get_spi_settings()));
+  status_ok_or_return(can_init(&s_can_storage, config_get_can_settings()));
+
+  status_ok_or_return(drv120_relay_init(config_get_drv120_relay_pin()));
+
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_initialize_sense_modules(SolarMpptCount mppt_count) {
+  status_ok_or_return(mppt_init());
+
+  status_ok_or_return(sense_init(config_get_sense_settings()));
+
+  SenseMcp3427Settings sense_mcp3427_settings;
+  status_ok_or_return(config_get_sense_mcp3427_settings(mppt_count, &sense_mcp3427_settings));
+  status_ok_or_return(sense_mcp3427_init(&sense_mcp3427_settings));
```

Comment:
The idea with the status_ok is that if the `mppt_count` is wrong, it'll return a status and not touch the settings pointer, so the status_ok returns gracefully instead of segfaulting trying to initialize garbage.

Reply:
You already check for nulls in the inits tho. It's a straight forward debug if you just return NULL for a bad count imo

Reply:
Hm fair enough, I'll change them to return const pointers